Cython/Build/Tests/TestRecythonize.py

```diff
@@ -0,0 +1,213 @@
+import shutil
+import os
+import tempfile
+
+import Cython.Build.Dependencies
+import Cython.Utils
+from Cython.TestUtils import CythonTest
+
+
+
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
+
+
+class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def сlear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
+
+    def setUp(self):
+        CythonTest.setUp(self)
+        self.сlear_function_and_Dependencies_caches()
+        self.temp_dir = (
+            tempfile.mkdtemp(
+                prefix='recythonize-test',
+                dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
+            )
+        )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
+
+    def tearDown(self):
+        CythonTest.tearDown(self)
+        self.сlear_function_and_Dependencies_caches()
+        shutil.rmtree(self.temp_dir)
+
+    def write_to_file(self, path, text):
+        with open(path, "w") as f:
+            f.write(text)
+
+    def fresh_cythonize(self, *args, **kwargs):
+        self.сlear_function_and_Dependencies_caches()
+        kwargs.update(language_level=self.language_level)
+        Cython.Build.Dependencies.cythonize(*args, **kwargs)
+
+    def refresh_dep_tree(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Utils.clear_method_caches(self.dep_tree)
+        self.dep_tree._transitive_cache.clear()
+
+    def fresh_all_dependencies(self, *args, **kwargs):
+        self.refresh_dep_tree()
+        return self.dep_tree.all_dependencies(*args, **kwargs)
+
+    def relative_lines(self, lines, line, start, end):
+        try:
+            ind = lines.index(line)
+            return lines[ind+start: ind+end]
+        except ValueError:
+            # XXX: It is assumed that VARS_LINE is always present.
+            ind = lines.index(VARS_LINE)
+            raise ValueError(
+                "{0!r} was not found, presumably in {1}".format(
+                    line, lines[ind-10: ind-1]))
```

Comment:
Should i replace `lines[ind-10: ind-1]` with `"".join(lines[ind-10: ind-1])`?

And therefore `{1}` to `{1!r}`